### PR TITLE
feat: Table component performance enhancement

### PR DESCRIPTION
**Pull request by thiagodallacqua-hpe**
## Description

Original ticket: DET-8110

## Test Plan

Go to `/det/projects/` and "play" with the table 😁 

## Commentary (optional)

There's a lot to uncover from the time that I've spent with this component...

I have chosen to take the path of understanding the component and why it's such a bottleneck for performance.

What I have found (in the react code) is that we have a very complex, highly customized table component with a lot of logic for resizing, re-organizing and filtering... all of which are tied to a global state (settings) which we know already that is kind of low on performance and has a few issues that we're trying to fix, plus we're actually getting to discuss replacing the architecture of it entirely (check @keita-determined's proposal).

The first issue found is that said global state has an auto-refresh which makes the whole tree re-renders, even if you don't do anything...

We're using `Context` to "avoid" prop drilling, but, we're using on a direct child, so I've removed it 😅.

There are a few listeners that keep changing the state, forcing re-renders, but, the state hasn't actually changed since the last update.

Resizing columns adds up to all of the above, constantly and indefinitely changing the state of the entire component tree.

very little code-reusability, mainly due to the complex nature of what we're trying to achieve.

I think that "CSS-ing" the resizing won't bring much to the table due to the fact that it relies on the event values to apply the new `width` to the `colgroup`, which, in turn reflects the `width` fo each `Cell` or `HeaderCell`, meaning that it is heavily dependant on the antd `Table`. It also uses table elements (although only to a certain point), which prevents the usage of `grid` or `flex-box`. (but, then again, it wouldn't help much anyway).

In conclusion, we can either manage the current state or shift to a new `Table` component 😅. 

For more insight, check PR's comments.

There's also some profiling json files attached to the ticket

Rendering sample before:

https://user-images.githubusercontent.com/104855841/191054936-103379e3-796a-4725-affc-03f4156c3053.mp4

after:

https://user-images.githubusercontent.com/104855841/191054996-7ee3d0c7-12a7-49cb-a039-48700f58f02c.mp4

resizing before:

https://user-images.githubusercontent.com/104855841/191079049-f6d1ebe8-2dd7-4704-ab37-9e72f470e6ac.mp4

resizing after:

https://user-images.githubusercontent.com/104855841/191079089-772b4af5-3f4b-41e4-add5-2e0856952619.mp4

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by keita-determined — DISMISSED** [review]
ColumnWidth of userSettings is not persistent.
Question: Did you also look into other table options?

https://user-images.githubusercontent.com/109113616/191137988-7e9cb354-eb14-482d-8379-07fb30e7293a.mov

File: webui/react/src/shared/configs/eslintrc.js
Comment (keita-determined) [inline webui/react/src/shared/configs/eslintrc.js:101]: could you fix this please?

**Review by trentwatt — DISMISSED** [review]
feels kinda buggy imo

https://user-images.githubusercontent.com/28383080/191338226-0cd61086-c8e3-43cb-b7f9-c7ee1033b6ce.mov

**Comment by keita-determined** [comment]
https://user-images.githubusercontent.com/109113616/191384095-1db6f754-10ce-4454-9ef8-e4fe3c73d46b.mov

Other trivial issue. 
Resizing bar doesn't stop when the bar exceed the minimum column size.

**Review by trentwatt — COMMENTED** [review]
> Resizing columns adds up to all of the above, constantly and indefinitely changing the state of the entire component tree.

the resizing causing a re-render is a newer development, probably caused by changes outside of the table component? I would rather we identify and fix that issue as opposed to adding another layer of things we would need to fix later. 

I will defer to alex on this one though.

**Comment by trentwatt** [comment]
> also, @trentwatt @keita-determined, as mentioned by @ClaireNeveu, although the resizing doesn't quite look perfect yet, it is way more performant than before, so, unless we really wanna brush it out here, or we disagree with the current code around it, can we merge this PR and fix the behavior in a new one?

I reached out to alex about it, I will ping him again.

File: webui/react/src/hooks/useSettings.ts
Comment (trentwatt) [inline webui/react/src/hooks/useSettings.ts:368]: keep

**Comment by ClaireNeveu** [comment]
If this is just waiting on design approval I'm gonna go ahead and approve it on behalf of design. Once the new designer is onboarded we'll revisit the table UX and focus on a more final solution.

**Pull request merged**
